test/integration/rmi_out_test.py

```diff
@@ -43,27 +51,146 @@ def test_deprish_out(rmi_out, request):
+    depreciation `data_source`.
+    """
+    if 'data_source' not in df:
+        df = pd.concat([
+            df.assign(data_source='PUC'),
+            df.assign(data_source='FERC'),
+        ])
+    return df
+
+
+@pytest.mark.parametrize(
+    "df1_name,df2_name,data_col,by_name", [
+        (
+            "grab_deprish",
+            "grab_deprish_to_eia",
+            'plant_balance_w_common',
+            'utilities'
+        ),
+        (
+            "grab_deprish",
+            "grab_deprish_to_eia",
+            'plant_balance_w_common',
+            'plants',
+        ),
+        (
+            "grab_deprish_to_eia",
+            "grab_deprish_to_ferc1",
+            'plant_balance_w_common',
+            'utilities',
+        ),
+        (
+            "grab_deprish_to_eia",
+            "grab_deprish_to_ferc1",
+            'plant_balance_w_common',
+            'plants',
+        ),
+        (
+            "grab_deprish",
+            "grab_deprish_to_ferc1",
+            'plant_balance_w_common',
+            'utilities'
+        ),
+        (
+            "grab_deprish",
+            "grab_deprish_to_ferc1",
+            'plant_balance_w_common',
+            'plants',
+        ),
+        (
+            "grab_ferc1_to_eia",
+            "grab_deprish_to_ferc1",
+            'capex_total',
+            'utilities',
+        ),
+        (
+            "grab_ferc1_to_eia",
+            "grab_deprish_to_ferc1",
+            'capex_total',
+            'plants',
+        ),
+    ]
+)
+def test_data_col_at_two_stages(
+    rmi_out,
+    df1_name: str,
+    df2_name: str,
+    data_col: str,
+    by_name: Literal['plants', 'utilities'],
+):
+    """
+    Test the consistency of a data column at two stages.
+
+    The data that is processed in this repo goes along multiple stages of its
+    journey. There are some quantities that are expected to be invariant when
+    aggregated at the plant level, and the utility level -- that the quantity
+    might slosh around between generators in a plant, or between plants within
+    a utility, but that when aggregated to the plant or utility level, it
+    should be the same at every step in the processing.
+
+    Ideally, the inputs
+    """
+    if by_name == 'plants':
+        by = PK_PLANTS
+    elif by_name == 'utilities':
+        by = PK_UTILS
+
+    test_stages = pudl_rmi.connect_deprish_to_ferc1.gb_test(
+        df1=rmi_out.__getattribute__(df1_name)().pipe(_add_data_source),
+        df2=rmi_out.__getattribute__(df2_name)().pipe(_add_data_source),
+        data_col=data_col,
+        by=by
+
+    )
+    # we add the off_rate.notnll() mask in here bc there are a ton of unconnected
+    # data in these tables (eg. depreciation records that don't have a EIA connection)
+    # these bbs will not get effectively propegated through each of these stages that
+    # require a match to EIA.
+    # there are enough of them that we want to generally ignore them
```

Comment:
Enough of what? What is "them" here?

Reply:
better? see also the test doc string

Reply:
Talking about "enough" makes it seems like there's some threshold that you have in mind, where you check sometimes but not others. I think stating that the test only checks for agreement when both data columns have non-null values is fine.

I still don't understand what `test_stages` means. It's data, not stages, right?

Reply:
With all the naming changes I think it might end up looking something like this:

```py
    test_df = agg_test_data(
        df1=rmi_out.__getattribute__(df1_name)().pipe(_add_data_source),
        df2=rmi_out.__getattribute__(df2_name)().pipe(_add_data_source),
        data_col=data_col,
        by=by,
    )
    actual_agg_errors = test_df[~test_df.data_isclose & test_df.data_ratio.notnull()]

    logger.info(
        f"Aggregation errors for {data_col} by {by_name} between {df1_name} "
        f"and {df2_name}: {len(actual_agg_errors)}"
    )
    expected_errors_path = (
        EXPECTED_ERRORS_PATH / f"expected_errors_{data_col}_{by_name}_{df1_name}_vs_"
        f"{df2_name}.csv"
    )
    expected_agg_errors = pd.read_csv(expected_errors_path).set_index(by)
    pd.testing.assert_index_equal(
        actual_agg_errors.index,
        expected_agg_errors.index,
        exact="equiv",
        check_order=False,
    )
```